File: clients/python/sky-island.py
# coding: utf-8

import json
import requests


class Encoder(json.JSONEncoder):
    def default(self, obj):
        if not isinstance(obj, Call):
            return super(Encoder, self).default(obj)
        return obj.__dict__


class Call(object):
    def __init__(self, url, call):
        self.url = url
        self.call = call


class Client(object):
    def __init__(self, host, port):
        self.endpoint = "http://{}:{}/api/v1/function".format(host, port)

    def function(self, url, call, full_resp=False):
        c = Call(url, call)
        res = requests.post(self.endpoint, json.dumps(c, cls=Encoder))
        if res.status_code == 200:
            data = json.loads(res.text)
            if full_resp:
                return data
            return data["data"]

File: clients/go/sky-island.go
package skyisland

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// Data holds the response from the API
type Data struct {
	Timestamp int64  `json:"timestamp"`
	Data      string `json:"data"`
}

// Client contains the HTTP client and the endpoint
// it needs to communicate with
type Client struct {
	hc       *http.Client
	endpoint string
}

// NewClient creates a new client usable with the Sky Island API
func NewClient(url string, port int, timeout time.Duration) *Client {
	c := &Client{
		hc: &http.Client{
			Timeout: timeout,
		},
		endpoint: fmt.Sprintf("%s:%d/api/v1/function", url, port),
	}
	return c
}

// Function makes the call to the API
func (c *Client) Function(url, call string) (*Data, error) {
	d := fmt.Sprintf(`{"url": "%s", "call": "%s"}`, url, call)
	req, err := http.NewRequest(http.MethodPost, c.endpoint, bytes.NewBuffer([]byte(d)))
	if err != nil {
		return nil, err
	}
	res, err := c.hc.Do(req)
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()
	var data Data
	if err := json.NewDecoder(res.Body).Decode(&data); err != nil {
		return nil, err
	}
	return &data, nil
}
